pkg/operator/configobservation/cloudprovider/observe_cloudprovider.go

```diff
@@ -3,45 +3,114 @@ package cloudprovider
 		if err := unstructured.SetNestedStringSlice(previouslyObservedConfig, currentCloudProvider, cloudProvidersPath...); err != nil {
 			errs = append(errs, err)
 		}
 	}
 
+	if len(existinCloudConfig) > 0 {
+		if err := unstructured.SetNestedStringSlice(previouslyObservedConfig, existinCloudConfig, cloudProviderConfPath...); err != nil {
+			errs = append(errs, err)
+		}
+	}
+
 	observedConfig := map[string]interface{}{}
 
 	infrastructure, err := listers.InfrastructureLister.Get("cluster")
 	if errors.IsNotFound(err) {
-		recorder.Warningf("ObserveRestrictedCIDRFailed", "Required infrastructures.%s/cluster not found", configv1.GroupName)
+		recorder.Warningf("ObserverCloudProviderNames", "Required infrastructures.%s/cluster not found", configv1.GroupName)
 		return observedConfig, errs
 	}
 	if err != nil {
 		return previouslyObservedConfig, errs
 	}
 
+	cloudProvider := getPlatformName(infrastructure.Status.Platform, recorder)
+	if len(cloudProvider) > 0 {
+		if err := unstructured.SetNestedStringSlice(observedConfig, []string{cloudProvider}, cloudProvidersPath...); err != nil {
+			errs = append(errs, err)
+		}
+	}
+
+	sourceCloudConfigMap := infrastructure.Spec.CloudConfig.Name
+	sourceCloudConfigNamespace := configNamespace
+	sourceLocation := resourcesynccontroller.ResourceLocation{
+		Namespace: sourceCloudConfigNamespace,
+		Name:      sourceCloudConfigMap,
+	}
+
+	if len(sourceCloudConfigMap) == 0 {
+		sourceLocation = resourcesynccontroller.ResourceLocation{}
+	}
+
+	err = listers.ResourceSyncer().SyncConfigMap(
+		resourcesynccontroller.ResourceLocation{
+			Namespace: targetNamespaceName,
+			Name:      "cloud-config",
+		},
+		sourceLocation)
+
+	if err != nil {
+		errs = append(errs, err)
+		return observedConfig, errs
+	}
+
+	if len(sourceCloudConfigMap) == 0 {
+		return observedConfig, errs
+	}
+
+	// usually key will be simply config but we should refer it just in case
+	staticCloudConfFile := fmt.Sprintf(cloudProviderConfFilePath, infrastructure.Spec.CloudConfig.Key)
+
+	if err := unstructured.SetNestedStringSlice(observedConfig, []string{staticCloudConfFile}, cloudProviderConfPath...); err != nil {
+		recorder.Warningf("ObserverCloudProviderNames", "Failed setting cloud-config : %v", err)
+		errs = append(errs, err)
+	}
+
+	if !equality.Semantic.DeepEqual(existinCloudConfig, []string{staticCloudConfFile}) {
+		recorder.Eventf("ObserverCloudProviderNamesChanges", "CloudProvider config file changed to %s", staticCloudConfFile)
+	}
+
+	return observedConfig, errs
+}
+
+func getPlatformName(platformType configv1.PlatformType, recorder events.Recorder) string {
 	cloudProvider := ""
-	switch infrastructure.Status.Platform {
+	switch platformType {
 	case "":
 		recorder.Warningf("ObserveCloudProvidersFailed", "Required status.platform field is not set in infrastructures.%s/cluster", configv1.GroupName)
-		return previouslyObservedConfig, errs
 	case configv1.AWSPlatform:
 		cloudProvider = "aws"
 	case configv1.VSpherePlatform:
```

Comment:
Can you please add support for Azure platform as well

Reply:
I can't add support for Azure in this PR. Because it was decided that applying a opaque configmap as a cloudprovider config will only be supported for vsphere for the time being. For other cloudproviders, plan is to validate and support only options we know work.